File: #0823.binary-trees-with-factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {

        int size = arr.size();
        std::sort(arr.begin(), arr.end());
        map<int,long> m;
        int mod = std::pow(10, 9) + 7;
        int result = 0;
        for (int i = 0; i < size; i++) {
            m[arr[i]]=1;
            for (int j = 0; j < i; ++j) {
                if (arr[i] % arr[j]==0) {
                    m[arr[i]] = (m[arr[i]] + m[arr[j]] * m[arr[i] / arr[j]]) % mod;
                }
            }
            result = (result + m[arr[i]]) % mod;
        }
        return result;
    }
};

File: #0823.binary-trees-with-factors.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var numFactoredBinaryTrees = function(arr) {
    arr.sort((left, right) => left - right);
    let m = new Map();
    let sol = Array(arr.length).fill(1);
    let mod = 1e9 + 7;
    arr.forEach((val) => {
        m.set(val, []);
    });

    for (let i = 0; i < arr.length; i++) {
        for (let j = i; j < arr.length; j++) {
            let prod = arr[i] * arr[j];
            if (prod > arr[arr.length - 1]) {
                break;
            }

            if (m.has(prod)) {
                m.get(prod).push([i, j]);
            }
        }
    }

    for (let i = 0; i < arr.length; i++) {
        if (m.has(arr[i])) {
            let prodArr = m.get(arr[i]);
            for (j = 0; j < prodArr.length; j++) {
                let i1 = prodArr[j][0], i2 = prodArr[j][1];
                sol[i] += sol[i1] * sol[i2] * (i1 == i2 ? 1 : 2) % mod;
            }
        }
    }

    return sol.reduce((sum, val) => (sum + val) % mod, 0);
};
